### PR TITLE
Operator Role

### PR DESCRIPTION
**Pull request by faze-quadrata**
## Checklist
1. [ ] Unit Testing?
2. [ ] Backward Compatible?
3. [ ] Upgrade Safe (storage)?
4. [ ] Updated Confluence?
5. [ ] Updated Gitbook?

### CODE REVIEW

File: contracts/storage/QuadConstant.sol [hunk 1]
```diff
@@ -6,6 +6,7 @@ contract QuadConstant {
     bytes32 public constant GOVERNANCE_ROLE = keccak256("GOVERNANCE_ROLE");
     bytes32 public constant READER_ROLE = keccak256("READER_ROLE");
     bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+    bytes32 public constant OPERATOR_ROLE = keccak256("OPERATOR_ROLE");
```
Comment (faze-quadrata) [inline contracts/storage/QuadConstant.sol:9]: Constants in Solidity are not stored in contract storage, but rather they are baked into the bytecode of the contract when it is compiled

Comment (faze-quadrata) [reply contracts/storage/QuadConstant.sol:9]: tested that these do not affect storage slots

```
contract QuadThrowaway {
  uint256 public someNum = 69; // slot 0x0
}
```

vs

```
contract QuadThrowaway {
  bytes32 public constant ISSUER_ROLE = keccak256("ISSUER_ROLE");
  uint256 public someNum = 69; // slot 0x0
  bytes32 public constant COW = keccak256("COW");
}
```

File: contracts/QuadGovernance.sol [hunk 2]
```diff
@@ -25,7 +25,10 @@ contract QuadGovernance is IQuadGovernance, AccessControlUpgradeable, UUPSUpgrad
         _setRoleAdmin(PAUSER_ROLE, GOVERNANCE_ROLE);
         _setRoleAdmin(ISSUER_ROLE, GOVERNANCE_ROLE);
         _setRoleAdmin(READER_ROLE, GOVERNANCE_ROLE);
+        _setRoleAdmin(OPERATOR_ROLE, GOVERNANCE_ROLE);
+
         _setupRole(GOVERNANCE_ROLE, _msgSender());
+        _setupRole(OPERATOR_ROLE, _msgSender());
```
Comment (fabrice-quadrata) [inline contracts/QuadGovernance.sol:31]: we don't want this for the sender.

File: contracts/storage/QuadConstant.sol (hunk 1 above)
Comment (fabrice-quadrata) [reply contracts/storage/QuadConstant.sol:9]: wooow, do you have a link for that?

File: utils/deployment.ts [hunk 3]
```diff
@@ -207,6 +208,16 @@ export const deployQuadrata = async (
     if (verbose)
       console.log(`[QuadGovernance] grant GOVERNANCE_ROLE to ${timelock}`);
   });
+
+  await recursiveRetry(async () => {
+    const tx = await governance.grantRole(OPERATOR_ROLE, timelock, {
```
Comment (fabrice-quadrata) [inline utils/deployment.ts:213]: the OPERATOR role should be assign to the Operator addresss. They are listed in Confluence.

Comment (fabrice-quadrata) [reply utils/deployment.ts:213]: you probably want this in the params, so you can reuse this script for unit tests

File: contracts/storage/QuadConstant.sol (hunk 1 above)
Comment (faze-quadrata) [reply contracts/storage/QuadConstant.sol:9]: https://ethereum.stackexchange.com/questions/140628/where-are-the-smart-contract-constants-stored

File: scripts/deployment/passport/deployMainnet.ts [hunk 4]
```diff
@@ -62,8 +72,13 @@ const {
     treasuryPerNetwork,
     multisigPerNetwork,
     TOKEN_IDS,
+    OPERATOR,
     true, // Verbose = true,
-    maxGasPerNetwork
+    maxGasPerNetwork,
+    "",
+    "",
```
Comment (fabrice-quadrata) [inline scripts/deployment/passport/deployMainnet.ts:79]: Isn't there 3 "" ?

Comment (fabrice-quadrata) [reply scripts/deployment/passport/deployMainnet.ts:79]: I would put `READER_ONLY` above the default value.

Comment (faze-quadrata) [reply scripts/deployment/passport/deployMainnet.ts:79]: no, just 2

File: utils/deployment.ts [hunk 5]
```diff
@@ -216,6 +219,15 @@ export const deployQuadrata = async (
...
 
+  await recursiveRetry(async () => {
+    const tx = await governance.grantRole(OPERATOR_ROLE, operator, {
+      maxFeePerGas,
+    });
+    await tx.wait();
+    if (verbose)
+      console.log(`[QuadGovernance] grant OPERATOR_ROLE to ${operator}`);
+  });
+
```
Comment (fabrice-quadrata) [inline utils/deployment.ts:230]: can we add the granting READER_ROLE?

File: scripts/deployment/passport/deployMainnet.ts (hunk 4 above)
Comment (fabrice-quadrata) [reply scripts/deployment/passport/deployMainnet.ts:79]: governanceAddress: string = "",
  passportAddress: string = "",
  readerAddress: string = "",


readerAddress is the QuadReader address

Comment (faze-quadrata) [reply scripts/deployment/passport/deployMainnet.ts:80]: yes

File: contracts/tests/TestQuadPassport.sol
Comment (fabrice-quadrata) [inline contracts/tests/TestQuadPassport.sol:1]: empty file

File: test/misc/test_storage_slot_value_contract_upgrade.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+      const TestQuadUpgrade = await ethers.getContractFactory(
+        "TestQuadUpgraded"
+      );
+      const upgradeContract = await upgrades.upgradeProxy(
+        originalContract.address,
+        TestQuadUpgrade,
+        { unsafeAllow: ["constructor"] }
+      );
+      expect(await upgradeContract.foo()).to.equal(2);
```
Comment (fabrice-quadrata) [inline test/misc/test_storage_slot_value_contract_upgrade.ts:30]: just for safety, can you also just add:

expect(await upgradeContract.someVal()).to.equal(69);
expect(await upgradeContract.someOtherVal()).to.equal(1337);

**Review by fabrice-quadrata — APPROVED** [review]

**Pull request merged**